From 74bf706f7d03150abf3455c8e55b88b65792db03 Mon Sep 17 00:00:00 2001
From: Tao Peng <jiankeng.pt@alibaba-inc.com>
Date: Tue, 4 Apr 2023 17:22:58 +0800
Subject: [PATCH] DeepRec diff.

---
 arc4random.c  | 10 ++++++++++
 evutil_rand.c |  5 -----
 2 files changed, 10 insertions(+), 5 deletions(-)

diff --git a/arc4random.c b/arc4random.c
index 8729f6b9..f0e9a7e8 100644
--- a/arc4random.c
+++ b/arc4random.c
@@ -45,6 +45,10 @@
 #define ARC4RANDOM_EXPORT
 #endif
 
+#include <stdint.h>
+#include "util-internal.h"
+#include "evthread-internal.h"
+
 #ifndef ARC4RANDOM_UINT32
 #define ARC4RANDOM_UINT32 uint32_t
 #endif
@@ -78,6 +82,12 @@
 /* Re-seed from the platform RNG after generating this many bytes. */
 #define BYTES_BEFORE_RESEED 1600000
 
+#define ARC4_LOCK_() EVLOCK_LOCK(arc4rand_lock, 0)
+#define ARC4_UNLOCK_() EVLOCK_UNLOCK(arc4rand_lock, 0)
+#ifndef EVENT__DISABLE_THREAD_SUPPORT
+static void *arc4rand_lock;
+#endif
+
 struct arc4_stream {
 	unsigned char i;
 	unsigned char j;
diff --git a/evutil_rand.c b/evutil_rand.c
index 8e9afdaa..2b1e1df6 100644
--- a/evutil_rand.c
+++ b/evutil_rand.c
@@ -120,11 +120,6 @@ ev_arc4random_buf(void *buf, size_t n)
 #define ssize_t EVENT__ssize_t
 #endif
 #define ARC4RANDOM_EXPORT static
-#define ARC4_LOCK_() EVLOCK_LOCK(arc4rand_lock, 0)
-#define ARC4_UNLOCK_() EVLOCK_UNLOCK(arc4rand_lock, 0)
-#ifndef EVENT__DISABLE_THREAD_SUPPORT
-static void *arc4rand_lock;
-#endif
 
 #define ARC4RANDOM_UINT32 ev_uint32_t
 #define ARC4RANDOM_NOSTIR
-- 
2.31.1

